File: typescript/src/ts/promptbuf.ts
import { JSONSchema7 } from "json-schema";

class PromptBuf {
	static encode(j: any): string {
		const oss: string[] = [];
		this.encodeValue(j, oss);
		return oss.join("");
	}

	static decode(minified: string, schema: JSONSchema7): any {
		const iss = minified.split("");
		return this.decodeValue(iss, schema);
	}

	private static encodeValue(j: any, oss: string[]): void {
		if (typeof j === "object" && j !== null) {
			if (Array.isArray(j)) {
				oss.push("{");
				for (let i = 0; i < j.length; i++) {
					this.encodeValue(j[i], oss);
					if (i < j.length - 1) oss.push(" ");
				}
				oss.push("}");
			} else {
				oss.push("[");
				const keys = Object.keys(j);
				for (let i = 0; i < keys.length; i++) {
					this.encodeValue(j[keys[i]], oss);
					if (i < keys.length - 1) oss.push(" ");
				}
				oss.push("]");
			}
		} else if (typeof j === "string") {
			oss.push(`"${j}"`);
		} else if (typeof j === "number") {
			oss.push(j.toString());
		} else if (typeof j === "boolean") {
			oss.push(j ? "1" : "0");
		} else if (j === null) {
			oss.push("null");
		}
	}

	private static decodeValue(iss: string[], schema: JSONSchema7): any {
		const c = iss.shift();

		if (c === "[") {
			return this.decodeObject(iss, schema);
		} else if (c === "{") {
			return this.decodeArray(iss, schema);
		} else if (c === '"') {
			return this.decodeString(iss);
		} else {
			iss.unshift(c!);
			return this.decodePrimitive(iss, schema);
		}
	}

	private static decodeObject(iss: string[], schema: JSONSchema7): any {
		const obj: any = {};
		const properties = schema.properties || {};
		const keys = Object.keys(properties);
		let i = 0;

		while (iss[0] !== "]" && i < keys.length) {
			obj[keys[i]] = this.decodeValue(iss, properties[keys[i]] as JSONSchema7);
			i++;
			if (iss[0] === ",") iss.shift();
		}
		iss.shift(); // consume ']'
		return obj;
	}

	private static decodeArray(iss: string[], schema: JSONSchema7): any[] {
		const arr: any[] = [];
		const elementSchema = Array.isArray(schema.items)
			? schema.items[0]
			: schema.items || {};

		while (iss[0] !== "}") {
			const nextChar = iss.shift()!;
			iss.unshift(nextChar);
			arr.push(this.decodePrimitive(iss, elementSchema as JSONSchema7));
		}
		iss.shift(); // consume '}'
		return arr;
	}

	private static decodeString(iss: string[]): string {
		let value = "";
		let char;
		while ((char = iss.shift()) !== '"') {
			value += char;
		}
		return value;
	}

	private static decodePrimitive(iss: string[], schema: JSONSchema7): any {
		const value = this.readUntil(iss, " }]");

		if (schema.type === "boolean") {
			return value === "1";
		} else if (schema.type === "integer") {
			return parseInt(value, 10);
		} else if (schema.type === "number") {
			return parseFloat(value);
		} else if (value === "null") {
			return null;
		} else {
			return value;
		}
	}

	private static readUntil(iss: string[], delimiters: string): string {
		let value = "";
		let char;
		while (iss.length > 0 && !delimiters.includes(iss[0])) {
			char = iss.shift();
			value += char;
		}
		return value;
	}
}

export default PromptBuf;

File: python/src/promptbuf.py
from typing import Any, Dict, List, Union

JSONSchema = Dict[str, Any]


def encode(v: Any, s: JSONSchema) -> str:
    output = []

    __encode(v, s, output)

    return " ".join(output)


def __encode(v: Any, s: JSONSchema, output: List[str]):
    s_type = s.get("type")

    if s_type == "object":
        ov = []
        if not s.get("properties"):
            return "{}"

        p_keys = list(s["properties"].keys())
        for key in p_keys:
            __encode(v.get(key), s["properties"][key], ov)

        output.append("{" + " ".join(ov) + "}")

    elif s_type == "array":
        if not s.get("items"):
            return "[]"
        av = []

        for item in v:
            __encode(item, s["items"], av)

        output.append("[" + " ".join(av) + "]")

    elif s_type == "string":
        if s.get("enum"):
            output.append(str(s["enum"].index(v)))
        else:
            output.append(f'"{v}"')

    elif s_type in ["integer", "number"]:
        output.append(str(v))

    elif s_type == "boolean":
        output.append("1" if v else "0")

    elif s_type == "null":
        output.append("null")

    else:
        return ""


def decode(v: str, s: JSONSchema) -> Any:
    return __decode(list(v), s)


def __decode(v: List[str], s: JSONSchema) -> Any:
    if not v:
        return None

    char = v.pop(0)

    if char == "{":
        return __decode_object(v, s)
    elif char == "[":
        return __decode_array(v, s)
    elif char == '"':
        return __decode_string(v)
    elif char in [" ", "}", "]"]:
        return __decode(v, s)
    else:
        v.insert(0, char)
        return __decode_primitive(v, s)


def __decode_object(v: List[str], s: JSONSchema) -> Dict[str, Any]:
    o = {}

    if not s.get("properties"):
        return None

    p_keys = list(s["properties"].keys())

    for key in p_keys:
        o[key] = __decode(v, s["properties"][key])

    return o


def __decode_array(v: List[str], s: JSONSchema) -> List[Any]:
    a = []
    while v[0] != "]":
        a.append(__decode(v, s["items"]))
    v.pop(0)  # Remove the closing bracket
    return a


def __decode_string(v: List[str]) -> str:
    value = read_until(v, ['"'])
    v.pop(0)  # Remove the closing quote
    return value


def __decode_primitive(
    v: List[str], s: JSONSchema
) -> Union[int, float, bool, None, str]:
    value = read_until(v, [" ", "}", "]"])

    s_type = s.get("type")

    if s_type == "integer":
        return int(value)
    elif s_type == "number":
        return float(value)
    elif s_type == "boolean":
        return value == "1"
    elif s_type == "null":
        return None
    elif s_type == "string":
        if s.get("enum"):
            return s["enum"][int(value)]
        return value


def read_until(v: List[str], delimiters: List[str]) -> str:
    value = ""

    while v and v[0] not in delimiters:
        value += v.pop(0)

    return value
